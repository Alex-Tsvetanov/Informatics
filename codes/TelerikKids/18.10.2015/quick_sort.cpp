#include <iostream>
#include <random>

using namespace std;

template <typename T>
T* get_pivot (T* a, T* b)
{
	T* masterPivot = a;
	T averrage = 0;
	for (T* it = a ; it != b ; it ++)
		averrage += (*it);
	averrage /= (b - a);
	for (T* it = a ; it != b ; it ++)
		if (abs ((*masterPivot) - averrage) > abs ((*it) - averrage))
			masterPivot = it;

	return masterPivot;
}

template <typename T>
T* fillArr (T* newArr, T* pivot, T* s, T* e)
{
	T* indOfPivot;
	size_t ind = 0;
	for (T* it = s ; it != e ; it ++)
		if ((*it) < (*pivot))
		{
			newArr [ind] = *it;
			ind ++;
		}
	indOfPivot = newArr + ind;
	for (T* it = s ; it != e ; it ++)
		if ((*it) >= (*pivot))
		{
			newArr [ind] = *it;
			ind ++;
		}
	return indOfPivot;
}

template <typename T>
void copyArr (T* newArr, T* s, T* e)
{
	T* newArrIt = newArr;
	for (T* it = s ; it != e ; it ++, newArrIt ++)
		*it = *newArrIt;
}

template <typename T>
void quick_sort (T* s, T* e)
{
	if (s + 1 == e)
		return ;
	if (s == e)
		return ;
	T* newArr = new T [e - s];
	T* pivot = fillArr (newArr, get_pivot (s, e), s, e);
	//for (T* it = s ; it != e ; it ++)
	//	cout << *it << " ";
	//cout << "\n" << *pivot << "\n----------------------------" << endl;
	quick_sort (newArr, pivot + 1);
	quick_sort (pivot + 1, newArr + (e - s));
	copyArr (newArr, s, e);
	delete[] newArr;
}

int main ()
{
	size_t SIZE;
	cin >> SIZE;
	int* arr = new int [SIZE];
	for (size_t i = 0 ; i < SIZE ; i ++)
		cin >> arr [i];
	quick_sort (arr + 0, arr + SIZE);
	for (size_t i = 0 ; i < SIZE ; i ++)
		cout << arr [i] << " ";
	cout << "\n";
}
