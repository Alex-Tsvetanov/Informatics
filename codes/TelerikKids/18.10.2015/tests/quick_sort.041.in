96
84 65 63 2 85 53 22 93 22 65 39 8 3 57 47 45 52 11 4 24 34 17 56 7 3 2 18 76 85 79 74 29 13 52 61 92 89 100 74 83 66 74 99 36 23 36 76 81 40 3 78 59 26 41 22 32 1 29 47 63 43 63 90 12 66 37 59 63 8 55 79 10 58 60 54 70 15 72 23 50 97 8 12 27 82 69 56 45 22 58 50 57 90 37 77 21 
