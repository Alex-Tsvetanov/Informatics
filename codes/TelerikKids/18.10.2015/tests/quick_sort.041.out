1 2 2 3 3 3 4 7 8 8 8 10 11 12 12 13 15 17 18 21 22 22 22 22 23 23 24 26 27 29 29 32 34 36 36 37 37 39 40 41 43 45 45 47 47 50 50 52 52 53 54 55 56 56 57 57 58 58 59 59 60 61 63 63 63 63 65 65 66 66 69 70 72 74 74 74 76 76 77 78 79 79 81 82 83 84 85 85 89 90 90 92 93 97 99 100 
